### PR TITLE
Add new 'config' syntax

### PR DESCRIPTION
**Pull request by jordalgo**
This new feature allows users to set environment variables and other configuration in bpftrace scripts. This makes scripts more portable e.g. it's easier to reproduce behaviors without having to remember how scripts were invoked.

Environment variables passed on the command line take precedence
over ones set in the script itself.

Example:
```
sudo bpftrace -v -e 'config = { BPFTRACE_STACK_MODE=raw; } kprobe:vfs_read {
print(kstack()); }'
```

##### Checklist

- [x] Language changes are updated in `man/adoc/bpftrace.adoc` and if needed in `docs/reference_guide.md`
- [x] User-visible and non-trivial changes updated in `CHANGELOG.md`
- [x] The new behaviour is covered by tests

### CODE REVIEW

**Comment by jordalgo** [comment]
I'm sure there are a bunch of things I didn't get right in this implementation but creating an RFC for the language syntax. I'll also be sure to add tests later on in the process.

cc @ajor

**Comment by viktormalik** [comment]
First of all, I think that this is a very useful feature which can improve portability of bpftrace scripts, thanks!

I didn't review the code at all, but have some thoughts on the syntax. Similarly to what @ajor [describes on iterators syntax](https://github.com/iovisor/bpftrace/issues/2775#issuecomment-1779899452), this breaks the traditional event-based concept of bpftrace probes.

I'd rather see initializing env variables in the `BEGIN` block (similarly to what awk does). To do that, we could introduce new syntax for environment variables, e.g. (inspired by CMake, we can come with something else):
```
BEGIN
{
    $ENV{BPFTRACE_STACK_MODE} = "raw";
}
```
The advantage of this is that we could eventually allow usage of environment variables inside other probes, too (and the value would be replaced by actual value upon compile time - similarly to positional parameters).

**Comment by jordalgo** [comment]
> this breaks the traditional event-based concept of bpftrace probes. I'd rather see initializing env variables in the BEGIN block

Yeah, I considered this too. I think I didn't do that initially because I worried about expectations around ordering in terms of when env variables get evaluated and set (in relation to evaluation of other things in `BEGIN`) but maybe that's not a concern (at least until we start allowing them to be used in place of other r-values in other probes). Anyway, I'm good with moving them to `BEGIN`.

As far as the syntax:
```
$ENV{BPFTRACE_STACK_MODE} = "raw";
```

Are you thinking we need `$ENV` prefix as opposed to just using `BPFTRACE_` so that we can support user defined env variables in the future? If so, what is the difference between them and global/scratch variables? Or is it just so we don't have to always prefix env variables with `BPFTRACE_`?

**Comment by viktormalik** [comment]
> Are you thinking we need `$ENV` prefix as opposed to just using `BPFTRACE_` so that we can support user defined env variables in the future? If so, what is the difference between them and global/scratch variables? Or is it just so we don't have to always prefix env variables with `BPFTRACE_`?

I'd say that they are quite different, both from user and from bpftrace point of view. As a user, I prefer to know which variables come from the environment and may therefore have unexpected values. As a bpftrace developer, it feels like parsing and handling env variables are different from normal variables (e.g. they can be used before their first assignment) and the prefix should simplify our work. Ofc if we stick with bpftrace-specific variables only (having the `BPFTRACE_` prefix), there's no need to add the `ENV` prefix.

**Comment by fbs** [comment]
I think putting everything in BEGIN will make BEGIN really confusing. #1945 has a proposal for CLI parsing, which would then also have to go in begin. 

If you go for something like `bpftrace:env:stackmode { "blaat" }`, then you dont need any special parsing tricks. Its just a standard probe, with a single value in it. IIRC the AST accepts that without modification. 
Because it starts with `bpftrace:` we can treat it as a special probe, with its own AST pass, with very little modifications outside the pass itself, we just need to filter it from other passes (could even do an ast rewrite for that).
I think it will also stick to the 'event based' probe idea, although its a special probe that fires once

**Comment by jordalgo** [comment]
> I think putting everything in BEGIN will make BEGIN really confusing.

Yeah, I'm starting to agree. I just don't like the inter-splicing of env var assignments with anything else in BEGIN which might imply some ordering when there is none; env vars will always get set before we do any codegen.

> `bpftrace:env:stackmode { "blaat" }`

This syntax feels a little confusing to me. I'd rather it look more like the actual env variable setting on the command line e.g. `env:BPFTRACE_STACK_MODE { perf }` or `bpftrace:env:BPFTRACE_STACK_MODE { perf }`  but TBH, I'm not crazy about the floating r-value in the brackets and prefer the original proposal `ENV { BPFTRACE_STACK_MODE=perf }`, which, if you look at the code in this PR, is not that invasive in terms of modifications - most of the business logic is in semantic_analyzer.cpp (though I'm not sure if that's the best place).

**Comment by viktormalik** [comment]
@fbs has a good point that the syntax should be unified with argument parsing and having everything in `BEGIN` would make it too complicated. I kind of start liking the idea of "meta" probes starting with `bpftrace:` prefix which are not executed but instead hold some script metadata. They seem quite intuitive, consistent with the current language, and quite extensible for future.

I agree with @jordalgo that we should use the original name - either `bpftrace:env:BPFTRACE_STACK_MODE { perf }` or, even better from my point of view,  `bpftrace:env { BPFTRACE_STACK_MODE = perf }`.

FWIW, I put this on the agenda of the next office hours meeting.

**Comment by viktormalik** [comment]
Also note that this has a big overlap with #728.

**Comment by fbs** [comment]
> > `bpftrace:env:stackmode { "blaat" }`
> 
> This syntax feels a little confusing to me. I'd rather it look more like the actual env variable setting on the command line e.g. `env:BPFTRACE_STACK_MODE { perf }` or `bpftrace:env:BPFTRACE_STACK_MODE { perf }` but TBH, I'm not crazy about the floating r-value in the brackets and prefer the original proposal `ENV { BPFTRACE_STACK_MODE=perf }`, which, if you look at the code in this PR, is not that invasive in terms of modifications - most of the business logic is in semantic_analyzer.cpp (though I'm not sure if that's the best place).

My main issue with this is that you now have a 3rd way to assign variables. So there will be `$var =x`, `@map = x` and `SPECIAL = x`. All the AST changes are due to this.
I agree that the `bpftraceprobe { val }` syntax is a bit weird too, my idea is that it should be read like `bpftrace:probe { return val }`, so the probe fires and returns a value. This is also a bit flawed as there is no const evaluation into a value.
Maybe there is another option that avoids creating new syntax?

The reason for me to go for `stack_mode` instead of `BPFTRACE_STACK_MODE` is because the latter is just shell conventions. bpftrace_ because env vars are not namespaced, this avoids conflict. uppercase because that is also normal for the env, probably avoid confusion with shell local variables? Either way, I'm not sure we should bring that shell detail into  bpftrace. Maybe we should even drop env from the name, why not just `bpftrace:config:stack_mode`?

**Comment by viktormalik** [comment]
> The reason for me to go for `stack_mode` instead of `BPFTRACE_STACK_MODE` is because the latter is just shell conventions. bpftrace_ because env vars are not namespaced, this avoids conflict. uppercase because that is also normal for the env, probably avoid confusion with shell local variables? Either way, I'm not sure we should bring that shell detail into bpftrace. Maybe we should even drop env from the name, why not just `bpftrace:config:stack_mode`?

If we define (and document) this as the idiomatic way of configuring bpftrace, then I'm all good with the last option. We can then just say that the configuration options can be also set via env variables which, by shell conventions, are capitalized and use the `BPFTRACE_` prefix.

**Comment by jordalgo** [comment]
Do we think there would ever be a difference between the `bpftrace:config` variables and env vars that you can set when invoking bpftrace? If not, then I'm not sure why you'd want to create a new thing that users would have to remember. I understand that these are shell conventions but is it so bad that we adopt them specifically so users don't have to remember which one needs a prefix and uppercase spelling? Somewhat rhetorical - I don't think it's so bad :)

To avoid another way of setting variables, what if we just re-used `$` e.g. `$BPFTRACE_` - it could be an AST change or we could just add branching logic for that ident prefix in semantic_analyzer.cpp.

**Comment by jordalgo** [comment]
Ok, I think this might be ready-ish. I'm sure there were a few things I probably missed.

File: src/ast/passes/semantic_analyser.cpp [hunk 1]
```diff
@@ -2672,6 +2672,222 @@ void SemanticAnalyser::visit(AssignVarStatement &assignment)
...
+    return false;
+  };
+
+  if (matches("BPFTRACE_STACK_MODE"))
+  {
+    if (!assignTy.IsStackModeTy())
+    {
+      invalidType();
+    }
+    else if (!env_is_set)
```
Comment (viktormalik) [inline src/ast/passes/semantic_analyser.cpp:2765]: TBH, I don't like that this check needs to be done for every config option. I can see the problem here (we'd need to parse env vars between doing semantic analysis of config and of the rest of the script) but I'm wondering if we can come up with some more clever and robust solution.

Idea: encapsulate config in a new class with various exported setters (for setting from env, script, cli, ...) which would automatically handle the priorities? It's a bit of work but I'm afraid that with the proposed approach, we will mess up the priorities sooner or later for some option.

File: src/ast/passes/semantic_analyser.cpp [hunk 2]
```diff
@@ -2672,6 +2672,222 @@ void SemanticAnalyser::visit(AssignVarStatement &assignment)
...
+  auto matches = [&](std::string &&env_var) -> bool {
+    if (env_ident == env_var)
+    {
+      env_is_set = std::getenv(env_var.c_str()) != nullptr;
+      return true;
+    }
+    return false;
+  };
+
+  if (matches("BPFTRACE_STACK_MODE"))
```
Comment (viktormalik) [inline src/ast/passes/semantic_analyser.cpp:2759]: This is probably super-nit but since we want to encourage moving from env vars to config options, we should not use the env var name rather than the config name (i.e. `"stack_mode"`) here. Also, if we introduce new config options without corresponding env vars inn future, the code will remain consistent.

File: src/bpftrace.h [hunk 3]
```diff
@@ -202,12 +213,8 @@ class BPFtrace
...
   bool resolve_user_symbols_ = true;
   bool debug_output_ = false;
-  enum class UserSymbolCacheType
-  {
-    per_pid,
-    per_program,
-    none,
-  } user_symbol_cache_type_;
+  bool verify_llvm_ir_ = false;
+  UserSymbolCacheType user_symbol_cache_type_;
```
Comment (viktormalik) [inline src/bpftrace.h:217]: It feels like introduction of these two fields could be moved into separate commit(s).

File: tests/runtime/config [hunk 4]
```diff
@@ -0,0 +1,25 @@
+NAME config as env var
```
Comment (viktormalik) [inline tests/runtime/config:1]: not env var, right?
```suggestion
NAME config
```

**Review by viktormalik — COMMENTED** [review]
Thanks for this work, I'm really looking forward to get it in! I've a couple of comments, though, see below.

Also, I'd love to see the work split into multiple commits, e.g. the refactoring of methods for setting various configs from env vars could be moved into separate commit(s).

**Review by danobi — COMMENTED** [review]
overall looks reasonable; haven't had a chance to look at details yet though.

one high level comment: could we put the config parsing (and future cli parsing) into a new AST pass? For stuff like AOT and @ajor's language plans, it would be useful to be able to selectively disable config/cli parsing via `PassManager`

File: tests/runtime/config (hunk 4 above)
Comment (jordalgo) [reply tests/runtime/config:1]: Maybe the name is confusing but I want to test that the env variable style e.g. `BPFTRACE_` works for setting the config (as opposed to the test below)

File: src/ast/passes/semantic_analyser.cpp (hunk 1 above)
Comment (jordalgo) [reply src/ast/passes/semantic_analyser.cpp:2765]: Yeah I don't like it either. Another class sounds like a good idea. Let me think on that in the context of @danobi's suggestion to move this parsing to a separate AST pass.

**Comment by jordalgo** [comment]
> Also, I'd love to see the work split into multiple commits, e.g. the refactoring of methods for setting various configs from env vars could be moved into separate commit(s).

Good suggestion. Let me work on that.

**Comment by jordalgo** [comment]
> could we put the config parsing (and future cli parsing) into a new AST pass?

Good call. Let me work on that too :)

File: tests/runtime/config (hunk 4 above)
Comment (viktormalik) [reply tests/runtime/config:1]: Ah, right, I just misunderstood it. I think that the current name is ok.

**Review by viktormalik — COMMENTED** [review]
Looks good to me, just one small nit.

Thanks for this work!

File: src/config.cpp [hunk 5]
```diff
@@ -34,16 +34,21 @@ Config::Config(bool has_cmd, bool bt_verbose) : bt_verbose_(bt_verbose)
...
-  else if (prevSource == ConfigSource::env_var)
+  else if (prevSource == source || prevSource == ConfigSource::env_var)
   {
     return false;
   }
+  else if (prevSource == ConfigSource::script &&
+           source == ConfigSource::env_var)
+  {
+    return true;
+  }
```
[the remark this answers is not in the document]
Comment (jordalgo) [reply src/config.cpp:51]: Good call, that's certainly cleaner. Will fix.

**Review by viktormalik — APPROVED** [review]

File: tests/config_analyser.cpp
[the remark this answers is not in the document]
Comment (ajor) [reply tests/config_analyser.cpp:1]: Also I think it would be good to check the error messages of failures, to ensure they stay reasonable. I started doing this in semantic analyser tests here: https://github.com/iovisor/bpftrace/blob/98960e19de63b670d753b77939dbdc78dff6ac15/tests/semantic_analyser.cpp#L2834

It'd be nice to start this early in this file before too many tests get added.

File: docs/reference_guide.md [hunk 6]
```diff
@@ -951,6 +955,30 @@ Attaching 1 probe...
...
+## 17. `config = {}`: Config Block
+
+To improve script portability, you can set bpftrace configs via the config block, 
+which can only be placed at the top of the script before any probes (even `BEGIN`).
+
+Example:
+```
+config = {
+    BPFTRACE_STACK_MODE=perf;
+    BPFTRACE_MAX_PROBES=2
```
Comment (ajor) [inline docs/reference_guide.md:967]:
```suggestion
    stack_mode=perf;
    max_probes=2
```

I'd prefer we make the non-env-var syntax the documented standard

**Review by ajor — CHANGES_REQUESTED** [review]
By adding these to the language syntax we're making them a lot more permanent than the env-vars were. I haven't thought too hard about the names yet (most might be fine!) but just want to suggest that we all consider what they're called. e.g. why is "max" at the beginning of "max_probes", yet at the end of "map_keys_max"?

On another note, I found some odd errors while playing around with this feature. Using non-literals results in a "none" type error:
```
$ sudo ./bpftrace -e 'config = { MAP_KEYS_MAX = $x } BEGIN { }'
stdin:1:25-26: ERROR: Invalid type for MAP_KEYS_MAX. Type: none                                                                                                       
config = { MAP_KEYS_MAX = $x } BEGIN { }                                                                                                                              
                        ~
```

Also for expressions which can be evaluated at compile-time:
```
$ sudo ./bpftrace -e 'config = { MAP_KEYS_MAX = 1 + 1 } BEGIN { }'                                           
stdin:1:25-26: ERROR: Invalid type for MAP_KEYS_MAX. Type: none                                                                                                       
config = { MAP_KEYS_MAX = 1 + 1 } BEGIN { }                                                                                                                           
                        ~
```

**Comment by jordalgo** [comment]
> e.g. why is "max" at the beginning of "max_probes", yet at the end of "map_keys_max"?

Yeah this bothered me too. Happy to change it (and cat_bytes_max) as there is a 3 to 2 majority for putting 'max' first. Should we just have backward support for BPFTRACE_MAP_KEYS_MAX or are we fine changing the env variable as well to BPFTRACE_MAX_MAP_KEYS (and MAX_CAT_BYTES) ?

> Using non-literals results in a "none" type error:

Good catch. Let me add a check for all non-literals. IMHO we shouldn't allow any other type (at least not in this first iteration).

File: src/ast/passes/semantic_analyser.cpp [hunk 7]
```diff
@@ -2685,6 +2682,23 @@ void SemanticAnalyser::visit(AssignVarStatement &assignment)
...
+{
+  assignment.expr->accept(*this);
+  auto &assignTy = assignment.expr->type;
+
+  if (is_final_pass())
+  {
+    auto &ty = assignTy.type;
+    if (ty == Type::none)
+    {
+      LOG(ERROR, assignment.loc, err_)
```
[the remark this answers is not in the document]
Comment (jordalgo) [reply src/ast/passes/semantic_analyser.cpp:2695]: Nope. Going to delete. Good catch.

**Comment by jordalgo** [comment]
Side question: Is there a reason why the language doesn't support boolean literals? Would be nice to do `debug_output=true` instead of `debug_output=1`

**Comment by ajor** [comment]
> Side question: Is there a reason why the language doesn't support boolean literals? Would be nice to do `debug_output=true` instead of `debug_output=1`

No reason in particular, it just hasn't come up before. Might be good to add at some point though.

File: src/config.h [hunk 8]
```diff
@@ -61,6 +64,25 @@ typedef std::variant<ConfigKeyBool,
                      ConfigKeyUserSymbolCacheType>
     ConfigKey;
 
+// These strings match the env variables (minus the 'BPFTRACE_' prefix)
+const std::map<std::string, ConfigKey> CONFIG_KEY_MAP = {
+  { "node_max", ConfigKeyInt::ast_max_nodes },
```
Comment (ajor) [inline src/config.h:69]: Similar to "verify_llvm_ir" - this one's just used for fuzzing: https://github.com/iovisor/bpftrace/pull/1601/commits/fb87a9f9f476af3d6cd6d3b42bea8bbd7235cc07
I'd say exclude.

>    This introduces `BPFTRACE_NODE_MAX` environment variable, and it
>    controls the maximum number of AST nodes. This feature is intended for
>    developers, and its usage is not included in the reference guide.

File: src/config.h [hunk 9]
```diff
@@ -61,6 +64,25 @@ typedef std::variant<ConfigKeyBool,
...
+  { "cat_bytes_max", ConfigKeyInt::cat_bytes_max },
+  { "debug_output", ConfigKeyBool::debug_output },
```
Comment (ajor) [inline src/config.h:71]: Also developer-only, so I'd say exclude

**Comment by ajor** [comment]
> > e.g. why is "max" at the beginning of "max_probes", yet at the end of "map_keys_max"?
> 
> Yeah this bothered me too. Happy to change it (and cat_bytes_max) as there is a 3 to 2 majority for putting 'max' first. Should we just have backward support for BPFTRACE_MAP_KEYS_MAX or are we fine changing the env variable as well to BPFTRACE_MAX_MAP_KEYS (and MAX_CAT_BYTES) ?

Yeah I'd vote for putting "max" first too. No strong opinion on changing the environment variables.

File: src/config.h [hunk 10]
```diff
@@ -61,6 +64,25 @@ typedef std::variant<ConfigKeyBool,
...
+  { "log_size", ConfigKeyInt::log_size },
+  { "map_keys_max", ConfigKeyInt::map_keys_max },
+  { "max_probes", ConfigKeyInt::max_probes },
+  { "max_bpf_progs", ConfigKeyInt::max_bpf_progs },
+  { "max_type_res_iterations", ConfigKeyInt::max_type_res_iterations },
+  { "no_cpp_demangle", ConfigKeyBool::no_cpp_demangle },
```
[the remark this answers is not in the document]
Comment (jordalgo) [reply src/config.h:77]: Ugh yes! I'm not a fan of negative named flags.

File: tests/config_analyser.cpp
[the remark this answers is not in the document]
Comment (jordalgo) [reply tests/config_analyser.cpp:1]: Added a few of both kinds of tests

**Comment by jordalgo** [comment]
@ajor I put the config and env var name changes into a separate commit. Let me know if you want me to merge it back into the first.

File: src/main.cpp [hunk 11]
```diff
@@ -363,6 +363,42 @@ static std::optional<struct timespec> get_delta_taitime()
...
+        << "BPFTRACE_STRLEN is deprecated. Use BPFTRACE_MAX_STRLEN instead.";
+    return false;
+  }
+
+  if (std::getenv("BPFTRACE_NODE_MAX"))
+  {
+    LOG(ERROR) << "BPFTRACE_NODE_MAX is deprecated. Use BPFTRACE_MAX_AST_NODES "
+                  "instead.";
+    return false;
+  }
```
[the remark this answers is not in the document]
Comment (jordalgo) [reply src/main.cpp:400]: That works for me. No strong feelings on this subject but I do like Strobelight.

File: src/main.cpp [hunk 12]
```diff
@@ -448,6 +450,7 @@ ast::PassManager CreateDynamicPM()
 ast::PassManager CreateAotPM()
 {
   ast::PassManager pm;
+  pm.AddPass(ast::CreateConfigPass());
```
[the remark this answers is not in the document]
Comment (jordalgo) [reply src/main.cpp:489]: Glad you called this out. I'm not very familiar with the AOT stuff so fine to drop it until we figure out how it'll work.

File: src/parser.yy [hunk 13]
```diff
@@ -349,6 +360,12 @@ stmt_list:
         |       block_stmt stmt_list    { $$ = $2; $2->insert($2->begin(), $1); }
         |       %empty                  { $$ = new ast::StatementList; }
                 ;
+                
+config_assign_stmt_list:
+                config_assign_stmt ";" config_assign_stmt_list { $$ = $3; $3->insert($3->begin(), $1); }
+        |       config_assign_stmt                             { $$ = new ast::StatementList; $$->push_back($1); }
+        |       %empty                                         { $$ = new ast::StatementList; }
```
[the remark this answers is not in the document]
Comment (jordalgo) [reply src/parser.yy:367]: Probably not very useful but bpftrace also supports empty BEGIN and END. No strong feeling either way.

[the remark this answers is not in the document]
Comment (danobi) [reply src/parser.yy:367]: Hmm, empty begin and end can still be used to check that probes attach correctly. But I suppose that could be said for empty config block too.

No strong feeling either way

[the remark this answers is not in the document]
Comment (jordalgo) [reply src/parser.yy:367]: Seems easy enough to get rid of. I'll try removing.

**Review by danobi — APPROVED** [review]
lgtm!

File: src/ast/passes/config_analyser.cpp [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+Pass CreateConfigPass()
+{
+  auto fn = [](Node &n, PassContext &ctx) {
+    auto configs = ConfigAnalyser(&n, ctx.b);
+    if (!configs.analyse())
+      return PassResult::Error("Config");
+    return PassResult::Success();
+  };
+
+  return Pass("Semantic", fn);
```
Comment (ajor) [inline src/ast/passes/config_analyser.cpp:201]:
```suggestion
  return Pass("ConfigAnalyser", fn);
```

**Review by ajor — COMMENTED** [review]
Just some copy-paste mistakes and nits left, looks good otherwise

File: src/main.cpp [hunk 15]
```diff
@@ -120,19 +121,19 @@ void usage()
   std::cerr << "ENVIRONMENT:" << std::endl;
   std::cerr << "    BPFTRACE_BTF                      [default: none] BTF file" << std::endl;
   std::cerr << "    BPFTRACE_CACHE_USER_SYMBOLS       [default: auto] enable user symbol cache" << std::endl;
-  std::cerr << "    BPFTRACE_CAT_BYTES_MAX            [default: 10k] maximum bytes read by cat builtin" << std::endl;
+  std::cerr << "    BPFTRACE_MAX_CAT_BYTES            [default: 10k] maximum bytes read by cat builtin" << std::endl;
```
[the remark this answers is not in the document]
Comment (jordalgo) [reply src/main.cpp:124]: Good call. Will do

File: docs/reference_guide.md [hunk 16]
```diff
@@ -170,19 +171,19 @@ OPTIONS:
 ENVIRONMENT:
     BPFTRACE_BTF                        [default: none] BTF file
     BPFTRACE_CACHE_USER_SYMBOLS         [default: (see docs)] enable user symbol cache
-    BPFTRACE_CAT_BYTES_MAX              [default: 10k] maximum bytes read by cat builtin
+    BPFTRACE_MAX_CAT_BYTES              [default: 10k] maximum bytes read by cat builtin
```
[the remark this answers is not in the document]
Comment (jordalgo) [reply docs/reference_guide.md:174]: There was a PR floating around here (https://github.com/iovisor/bpftrace/pull/2754) to just use the man page and delete the reference guide but I think it got abandoned :(

File: src/ast/passes/config_analyser.cpp (hunk 14 above)
Comment (jordalgo) [reply src/ast/passes/config_analyser.cpp:201]: Goodc catch!

File: src/main.cpp [hunk 17]
```diff
@@ -362,6 +370,35 @@ static std::optional<struct timespec> get_delta_taitime()
...
   }
 
+  if (!get_bool_env_var("BPFTRACE_NO_CPP_DEMANGLE", [&](bool x) {
+        LOG(WARNING) << "BPFTRACE_NO_CPP_DEMANGLE is deprecated. Use "
+                        "BPFTRACE_CPP_DEMANGLE=0 instead.";
+        config_setter.set(ConfigKeyBool::cpp_demangle, !x);
+      }))
+    return false;
+
+  if (!get_uint64_env_var("BPFTRACE_MAX_CAT_BYTES", [&](uint64_t x) {
```
[the remark this answers is not in the document]
Comment (jordalgo) [reply src/main.cpp:380]: Doh. Stupid autocomplete :) Good catch.

**Comment by jordalgo** [comment]
@ajor Thanks for the sharp-eyed review. I think I fixed all that you mentioned.

Additionally, I added one more commit whereby I extract out all the env-only configs from the config class. There were many other environment variables that weren't included in the config (e.g. BPFTRACE_BTF, BPFTRACE_VMLINUX, etc...) so
it was either include them in the config class or extract out similar ones. I chose the latter as it makes more sense to only include config vars that can be set by both the env and the config syntax (less indirection and clearer delineation).

**Review by ajor — APPROVED** [review]

**Comment by jordalgo** [comment]
@viktormalik @danobi @ajor - Thanks so much for the great reviews and guidance on this one!
